src/v/coproc/wasm_event_listener.cc

```diff
@@ -0,0 +1,202 @@
+            .then([this](std::vector<wasm_script_action> wsas) {
+                return ss::do_with(
+                  std::move(wsas),
+                  [this](std::vector<wasm_script_action>& wsas) {
+                      return ss::do_for_each(
+                        wsas, [this](wasm_script_action& wsa) {
+                            /// Finally the action is performed, either the
+                            /// source is written to disk or a file is deleted
+                            /// from disk
+                            return resolve_wasm_script(std::move(wsa));
+                        });
+                  });
+            })
+            .then([this] {
+                /// Every 2s the loop will pause, always looping attempting to
+                /// deploy or remove the next coprocessor script
+                return ss::sleep_abortable(2s, _abort_source)
+                  .handle_exception([](std::exception_ptr eptr) {
+                      vlog(
+                        coproclog.debug, "Exited sleep call early: {}", eptr);
+                  });
+            });
+      });
+}
+
+/// Map a model::record_batch -> std::vector<wasm_script_action>
+/// The transform will only occur after the record has passed all validators
+std::vector<wasm_script_action> wasm_event_listener::process_wasm_events(
+  model::record_batch_reader::data_t& events) {
+    std::vector<wasm_script_action> wsas;
+    for (auto& record_batch : events) {
+        record_batch.for_each_record([&wsas](model::record r) {
+            wasm_event_errc mb_error = wasm_event_validate(r);
+            if (mb_error != wasm_event_errc::none) {
+                vlog(
+                  coproclog.error,
+                  "Erranous coproc record detected, issue: {}",
+                  mb_error);
+                return;
+            }
+            auto id = wasm_event_get_id(r);
+            wsas.emplace_back(wasm_script_action{
+              .name = std::move(*id), .source_code = r.share_value()});
+        });
+    }
+    return wsas;
+}
+
+ss::future<ss::stop_iteration>
+wasm_event_listener::poll_topic(model::record_batch_reader::data_t& events) {
+    return _client.fetch_partition(_coproc_internal_tp, _offset, 64_KiB, 5s)
+      .then([this, &events](kafka::fetch_response response) {
+          if (
+            (response.error != kafka::error_code::none)
+            || _abort_source.abort_requested()) {
+              return ss::stop_iteration::yes;
+          }
+          vassert(response.partitions.size() == 1, "Unexpected partition size");
+          auto& p = response.partitions[0];
+          vassert(p.name == _coproc_internal_tp.topic, "Unexpected topic name");
+          vassert(p.responses.size() == 1, "Unexpected responses size");
+          auto& pr = p.responses[0];
+          if (!pr.has_error()) {
+              iobuf record_set(std::move(*pr.record_set));
+              while (!record_set.empty()) {
```

Comment:
Nice. Looks like it will be useful to subsume this pattern into the fetch api. I have a half-considered `record_set` type for helping here.

Reply:
Ok